File: prsense-node/src/queue/queue.config.ts
// src/queue/QueueHandlerRegistry.ts

import { QueueEventHandler } from "../services/QueueEventHandler";
import { EventName, EventNameForPush } from "../constants/EventNames";
import { Config } from "../config";
import AWS from 'aws-sdk';

type QueueHandler = {
  instance: any;
  queueName: string;
  callback: (instance: any) => (msg: AWS.SQS.Message) => Promise<void>;
};

export class QueueHandlerRegistry {
  private static _instance: QueueHandlerRegistry;
  private queueEventHandler = QueueEventHandler.getInstance();

  private handlers: Record<EventName, QueueHandler> = {
    [EventName.EMBEDDING_RESULT]: {
      instance: this.queueEventHandler,
      queueName: Config.NODEJS_SERVICE_QUEUE,
      callback: (i) => i.handleEmbeddingResult,
    },
    // [EventName.CHUNK]: {
    //   instance: this.queueEventHandler,
    //   queueName: Config.NODEJS_SERVICE_QUEUE,
    //   callback: (i) => i.handleEmbeddingResult,
    // }
  };

  private pushConfig: Record<EventNameForPush, string> = {
    [EventNameForPush.CHUNK]: Config.NODEJS_SERVICE_QUEUE,
    [EventNameForPush.ANALYZE_PR]: Config.NODEJS_SERVICE_QUEUE,
  };

  static getInstance(): QueueHandlerRegistry {
    if (!this._instance) {
      this._instance = new QueueHandlerRegistry();
    }
    return this._instance;
  }

  getHandlers(): Record<EventName, QueueHandler> {
    return this.handlers;
  }

  getPushQueue(event: EventNameForPush): string {
    return this.pushConfig[event];
  }
}

File: prsense-ai/app/queues/queue_config.py
from config import EnvConfig
from constants.event_name import EventName, EventNameForPush
from services.queue_event_handler import QueueEventHandler

class Config:

    _instance = None

    def __call__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super(Config).__call__(*args, **kwargs)
        return cls._instance
    
    def __init__(self):

        if hasattr(self, "_initialized") and self._initialized:
            return
        
        self.queueEventHandler = QueueEventHandler(self)
        
        self.pushConfig = {
            EventNameForPush.EMBEDDING_RESULT.value: EnvConfig.PYTHON_SERVICE_QUEUE,
            EventName.CHUNK.value: EnvConfig.NODE_SERVICE_QUEUE,
        }

        self.listenConfig = {
            EventName.CHUNK.value: self.queueEventHandler.handle_chunk,
        }

    def get_queue_name(self, event_name):
        return self.pushConfig.get(event_name)
    
    def get_handler(self, event_name):
        return self.listenConfig.get(event_name) 
